sdk/clientcore/core/src/main/java/io/clientcore/core/util/Union.java

```diff
@@ -0,0 +1,340 @@
+        }
+
+        ArrayList<Type> typeCopy = new ArrayList<>(types.length);
+        for (int i = 0; i < types.length; i++) {
+            final Type currentType = types[i];
+            if (currentType == null) {
+                throw LOGGER.logThrowableAsError(
+                    new IllegalArgumentException("types cannot contain null values: null value in index " + i));
+            } else if (!(currentType instanceof Class<?> || currentType instanceof ParameterizedType)) {
+                throw LOGGER.logThrowableAsError(new IllegalArgumentException(
+                    String.format("types must be of type Class or ParameterizedType: type name is %s in index %d.",
+                        currentType.getTypeName(), i)));
+            }
+
+            typeCopy.add(types[i]);
+        }
+        this.types = typeCopy;
+    }
+
+    /**
+     * Creates a new instance of {@link Union} with the provided types.
+     * <p>
+     * Currently, the types can be of type {@link Class} or {@link ParameterizedType}. If the type is a {@link Class},
+     * it represents a simple type. If the type is a {@link ParameterizedType}, it represents a generic type.
+     * For example, {@code List<String>} would be represented as {@code new GenericParameterizedType(List.class, String.class)}.
+     * </p>
+     *
+     * It throws {@link IllegalArgumentException} if:
+     * <ul>
+     *   <li>value is not of one of the types in the union,</li>
+     *   <li>value is null,</li>
+     *   <li>types array is null or empty,</li>
+     *   <li>types array contains a null value,</li>
+     *   <li>types array contains a type that is not of type {@link Class} or {@link ParameterizedType}.</li>
+     * </ul>
+     *
+     * @param types The types of the union.
+     * @return A new instance of {@link Union}.
+     */
+    public static Union ofTypes(Type... types) {
+        return new Union(types);
+    }
+
+    /**
+     * Sets the value of the union.
+     *
+     * @param value The value of the union.
+     * @throws IllegalArgumentException If the value is not of one of the types in the union.
+     */
+    @SuppressWarnings("unchecked")
+    public void setValue(Object value) {
+        for (Type type : types) {
+            if (isInstanceOfType(value, type) || isPrimitiveTypeMatch(value, type)) {
+                this.value = value;
+                this.currentType = type;
+                return;
+            }
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Invalid type: " + value.getClass().getName()));
+    }
+
+    /**
+     * Gets the type of the value.
+     *
+     * @return The type of the value.
+     */
+    public Type getCurrentType() {
+        return currentType;
+    }
+
+    /**
+     * Gets the types of the union.
+     *
+     * @return The types of the union.
+     */
+    public List<Type> getSupportedTypes() {
+        return Collections.unmodifiableList(types);
```

Comment:
We should just create this `unmodifiableList` once and not recreate on each invocation of this method